File: LevelAndroidSDK/sdk/src/main/java/com/theshopatvsp/levelandroidsdk/ble/model/DeviceStateMachine.java
package com.theshopatvsp.levelandroidsdk.ble.model;

import android.content.Intent;
import android.util.Log;

import  com.theshopatvsp.levelandroidsdk.ble.model.constants.BleDeviceOutput;
import  com.theshopatvsp.levelandroidsdk.ble.model.constants.reporter.DependentDataScale;
import  com.theshopatvsp.levelandroidsdk.ble.model.constants.reporter.DependentDataType;
import  com.theshopatvsp.levelandroidsdk.ble.model.constants.reporter.DependentVariableDescription;
import  com.theshopatvsp.levelandroidsdk.ble.model.constants.reporter.IndependentVariableDescription;
import  com.theshopatvsp.levelandroidsdk.ble.model.constants.state.DeviceInteractionState;
import  com.theshopatvsp.levelandroidsdk.ble.model.response.CodePacket;
import  com.theshopatvsp.levelandroidsdk.ble.model.response.DataPacket;
import  com.theshopatvsp.levelandroidsdk.ble.model.response.LockPacket;
import  com.theshopatvsp.levelandroidsdk.ble.model.response.ReportAttributesData;
import  com.theshopatvsp.levelandroidsdk.ble.model.response.TimePacket;
import  com.theshopatvsp.levelandroidsdk.ble.model.response.TransmitControlData;

/**
 * Created by andrco on 6/12/16.
 */
public class DeviceStateMachine {
    private static final String TAG = DeviceStateMachine.class.getSimpleName();
    private DeviceInteractionState state;
    private boolean reporter0Right = false, reporter1Right = false, reporter0IsOn = false, reporter1IsOn = false;
    private boolean reporter2Right = false, reporter2IsOn = false;
    private int packetsToDownload = 0, currentPacketsDownloaded = 0;
    private boolean timeIsCorrect = false, transmitControlOn = false, needLedCode = false;
    private long timeDiff, deviceTime;


    private ReportAttributesData reporter0 = new ReportAttributesData(0, IndependentVariableDescription.SECONDS, 15,
            DependentVariableDescription.STEP_PER_TIME, DependentDataType.UINT8, DependentDataScale.ONE_TO_ONE_BIT,
            0, 4, 0);
    private ReportAttributesData reporter1 = new ReportAttributesData(1, IndependentVariableDescription.ON_CHANGE, 1,
            DependentVariableDescription.BATTERY_PERCENT_REMAIN, DependentDataType.INT24, DependentDataScale.ONE_TO_ONE_BIT,
            0, 1, 0);
    private ReportAttributesData reporter2 = null;
    public DeviceStateMachine() {
    state = DeviceInteractionState.QueryLock;
    }

    public DataPacket processResult(DataPacket data) {
        Log.v(TAG, "processing result state = " + state);

        if (state == DeviceInteractionState.QueryLock && data instanceof LockPacket) {
            LockPacket lock = (LockPacket)data;

            if (lock.getLock() == 0) {
                needLedCode = true;
            } else {
                needLedCode = false;
                state = DeviceInteractionState.QueryTime;
                return null;
            }
        }

        if (state.name().toLowerCase().startsWith("sendledcode") && data instanceof CodePacket) {
            state = state.success();
            return null;
        }

        /*if (state == DeviceInteractionState.DownloadData && currentPacketsDownloaded < packetsToDownload) {
            Log.v(TAG, "Downloading data total = " + packetsToDownload + " current = " + currentPacketsDownloaded);
            return null;
        }*/

        if( state == DeviceInteractionState.QueryTime && data != null && data instanceof TimePacket ) {
            TimePacket time = (TimePacket) data;

            deviceTime = time.getTimestamp();

            if( Math.abs(time.getReceived().getTime() - time.getTimestamp()) < (60 * 1000) ) {
                timeIsCorrect = true;
            } else {
                timeDiff = time.getReceived().getTime() - time.getTimestamp();
            }
        }

        /*if( state == DeviceInteractionState.QueryReporter0 && data != null && data.equals(DeviceInteractionState.SetupReporter0.getPacket(0))) {
            Log.v(TAG, "reporter 0 is right!!!");
            reporter0Right = true;
        }

        if( state == DeviceInteractionState.QueryReporter1 && data != null && data.equals(DeviceInteractionState.SetupReporter1.getPacket(0))) {
            Log.v(TAG, "reporter 1 is right!!!");
            reporter1Right = true;
        }

        if( state == DeviceInteractionState.QueryReporter2 && data != null && data.equals(DeviceInteractionState.SetupReporter2.getPacket(0))) {
            Log.v(TAG, "reporter 2 is right!!!");
            reporter2Right = true;
        }

        if (state == DeviceInteractionState.QueryReportControl && data != null) {
            int reportControl = data.getReportControl();

            if ((reportControl & 0x01) == 0x01) {
                reporter0IsOn = true;
            }

            if ((reportControl & 0x02) == 0x02) {
                reporter1IsOn = true;
            }

            if ((reportControl & 0x04) == 0x04) {
                reporter2IsOn = true;
            }
        }

        if (state == DeviceInteractionState.QueryData && data instanceof TransmitControlData) {
            packetsToDownload = ((TransmitControlData)data).getTotalRecordCount();
        }

        if (state == DeviceInteractionState.DrainData) {
            transmitControlOn = true;
        }*/

        state = state.success();

        int reporters = 0;

        if (state == DeviceInteractionState.SetTime && timeIsCorrect) {
            state = state.success();
        }

        /*if( state == DeviceInteractionState.DisableReporters && ((reporter0Right && reporter1Right & reporter2IsOn) || data.getReportControl() == 0) ) {
            state = state.success();
        }

        if( state == DeviceInteractionState.DisableReporters ) {
            if( reporter0Right ) {
                reporters |= 0x01;
            } else {
                reporter0IsOn = false;
            }

            if( reporter1Right ) {
                reporters |= 0x02;
            } else {
                reporter1IsOn = false;
            }

            if (reporter2Right) {
                reporters |= 0x04;
            } else {
                reporter2IsOn = false;
            }
        }

        if (state == DeviceInteractionState.DrainData && packetsToDownload == 0) {
            state = state.success();
            state = state.success();
        }

        if( state == DeviceInteractionState.SetupReporter0 && reporter0Right ) {
            Log.v(TAG, "not setting reporter 0 cuz its right!!!");
            state = state.success();
        }

        if( state == DeviceInteractionState.SetupReporter1 && reporter1Right ) {
            Log.v(TAG, "not setting reporter 1 cuz its right!!!");
            state = state.success();
        }

        if( state == DeviceInteractionState.SetupReporter2 && reporter2Right ) {
            Log.v(TAG, "not setting reporter 2 cuz its right!!!");
            state = state.success();
        }

        if( state == DeviceInteractionState.EnableReporters && reporter0IsOn && reporter1IsOn && reporter2IsOn ) {
            state = state.success();
        }

        if (state == DeviceInteractionState.EnsureTransmitControlIsOn && transmitControlOn) {
            state = state.success();
        }*/

        Log.v(TAG, "processing result OUT state = " + state);

        return state.getPacket(reporters);
    }

    public Intent disconnect() {
        state = state.uhoh();

        if( state.name().toLowerCase().startsWith("sendledcode") ) {
            Log.v(TAG, "disconnected, sending LedCodeFailed");
            return new Intent(BleDeviceOutput.LedCodeFailed.name());
        }

        reporter0Right = false;
        reporter1Right = false;
        reporter0IsOn = false;
        reporter1IsOn = false;
        reporter2Right = false;
        reporter2IsOn = false;
        packetsToDownload = 0;
        currentPacketsDownloaded = 0;
        timeIsCorrect = false;
        transmitControlOn = false;
        timeDiff = 0;

        return null;
    }

    public void isBonded() {
        state = DeviceInteractionState.QueryLock;
    }

    public DeviceInteractionState getState() {
        return state;
    }

    public boolean sendPacket(int reporter) {
        if ((reporter == 0 && reporter0Right) || (reporter == 1 && reporter1Right) ||
                (reporter == 2 && reporter2Right)) {
            return true;
        }

        return false;
    }

    public boolean isTimeIsCorrect() {
        return timeIsCorrect;
    }

    public long getTimeDiff() {
        return timeDiff;
    }

    public void incPacketToDownload() {
        currentPacketsDownloaded++;
    }

    public boolean needLedCode() {
        return needLedCode;
    }

    public long getDeviceTime() {
        return deviceTime;
    }

    public void reset() {
        reporter0Right = false;
        reporter1Right = false;
        reporter0IsOn = false;
        reporter1IsOn = false;
        reporter2Right = false;
        reporter2IsOn = false;
        packetsToDownload = 0;
        currentPacketsDownloaded = 0;
        timeIsCorrect = false;
        transmitControlOn = false;
        timeDiff = 0;
        deviceTime = 0;
        if( state == DeviceInteractionState.Done ) {
            state = DeviceInteractionState.QueryTime;
        } else {
            state = DeviceInteractionState.QueryLock;
        }
    }
}

File: LevelIOSSdkExample/LevelIOSSdkExample/LevelIOSSdk/DeviceStateMachine.swift
//
//  DeviceStateMachine.swift
//  LevelIOSSDK
//
//  Created by Andrew Cook on 6/29/16.
//  Copyright © 2016 TheShop. All rights reserved.
//

import Foundation

enum DeviceLifecycle: String {
    case QueryLock, SendLedCode1, SendLedCode2, SendLedCode3, SendLedCode4, QueryTime,
        SetTime, QueryReportControl, QueryReporter0, QueryReporter1, QueryReporter2, Done, Unknown
    
    func getCommand() -> DeviceCommand {
        switch self {
        case .QueryTime: return DeviceCommand.TimeRD
        case .SetTime: return DeviceCommand.TimeWR
        case .QueryReportControl: return DeviceCommand.ReportControl
        case .QueryReporter0, .QueryReporter1, .QueryReporter2: return DeviceCommand.ReporterAttributes
        case .QueryLock: return DeviceCommand.LockRD
        default: return DeviceCommand.Unknown
        }
    }
    
    func getPacket() -> [UInt8] {
        switch self {
        case .SetTime:
            return BitsHelper.convertTo4Bytes(time: NSDate().timeIntervalSince1970)
        case .QueryReporter0:
            let bytes: [UInt8] = [0]
            return bytes
        case .QueryReporter1:
            let bytes: [UInt8] = [1]
            return bytes
        case .QueryReporter2:
            let bytes: [UInt8] = [2]
            return bytes
        default: return [UInt8]()
        }
    }
    
    func getTransition() -> DeviceTransition {
        switch self {
        case .SendLedCode1, .SendLedCode2, .SendLedCode3, .SendLedCode4:
            return .CodeWasReceived
        case .QueryTime:
            return .GotTime
        case .SetTime:
            return .TimeWasSet
        case .QueryReportControl:
            return .GotReporterControl
        case .QueryReporter0:
            return .GotReporter0Attriburtes
        case .QueryReporter1:
            return .GotReporter1Attriburtes
        case .QueryReporter2:
            return .GotReporter2Attriburtes
        default:
            return .Unknown
        }
    }
}

enum DeviceTransition {
    case GotLock, NeedLedCode, GotTime, Disconnected, CodeWasReceived, TimeWasSet, GotReporterControl, GotReporter0Attriburtes, GotReporter1Attriburtes, GotReporter2Attriburtes,
    ReportersDisabled, GotTransmitControl, TurnedTransmitControlOn, GotData, Reporter0AttributesSet, Reporter1AttributesSet, Reporter2AttributesSet, ReporterEnabled, Unknown
}

class DeviceStateMachine {
    var stateMachine: StateMachine<DeviceLifecycle, DeviceTransition>
    var currentTransition: Int = 0
    var enabled: Bool = false
    var reporter0Config: ReportAttributes? = nil, reporter1Config: ReportAttributes? = nil, reporter2Config: ReportAttributes? = nil
    var reporter0On: Bool = false, reporter1On: Bool = false, reporter2On: Bool = false
    var timeIsCorrect: Bool = false, transmitControlOn = false
    var timeDiff: Double = 0, deviceTime: Double = 0
    var totalRecordsToDownload: Int = 0
    var recordsDownloaded: Int = 0
    var ledCodeNeeded: Bool = false
    
    var expectedReport0Attributes: ReportAttributes = ReportAttributes(reporter: 0, indVarDesc: IndependentVariableDescription.Seconds, indVarScale: 15, depVarDesc: DependentVariableDescription.StepPerTime,
                                                                      depDataType: DependentDataType.UInt8, depDataScale: DependentDataScale.OneToOneBit, dataFieldsPerSample: 0, samplesPerRecord: 4, maxRecordsPerReport: 0)
    var expectedReport1Attributes: ReportAttributes = ReportAttributes(reporter: 1, indVarDesc: IndependentVariableDescription.OnChange, indVarScale: 1, depVarDesc: DependentVariableDescription.BatteryPercentRemaining,
                                                                       depDataType: DependentDataType.Int24, depDataScale: DependentDataScale.OneToOneBit, dataFieldsPerSample: 0, samplesPerRecord: 1, maxRecordsPerReport: 0)
    var expectedReport2Attributes: ReportAttributes = ReportAttributes(reporter: 2, indVarDesc: IndependentVariableDescription.Seconds, indVarScale: 5, depVarDesc: DependentVariableDescription.AccelerometerFILT,
                                                                       depDataType: DependentDataType.Int16, depDataScale: DependentDataScale.PlusMinus1GToDataSize, dataFieldsPerSample: 0, samplesPerRecord: 24, maxRecordsPerReport: 0)
    
    init() {
        stateMachine = StateMachine<DeviceLifecycle, DeviceTransition>(state: .QueryLock)
        
        stateMachine.addTransition(transition: .NeedLedCode, from: .QueryLock, to: .SendLedCode1)
        stateMachine.addTransition(transition: .GotLock, from: .QueryLock, to: .QueryTime)
        stateMachine.addTransition(transition: .GotTime, from: .QueryTime, to: .SetTime)
        stateMachine.addTransition(transition: .CodeWasReceived, from: .SendLedCode1, to: .SendLedCode2)
        stateMachine.addTransition(transition: .CodeWasReceived, from: .SendLedCode2, to: .SendLedCode3)
        stateMachine.addTransition(transition: .CodeWasReceived, from: .SendLedCode3, to: .SendLedCode4)
        stateMachine.addTransition(transition: .CodeWasReceived, from: .SendLedCode4, to: .QueryTime)
        stateMachine.addTransition(transition: .Disconnected, from: .SendLedCode4, to: .QueryLock)
        stateMachine.addTransition(transition: .TimeWasSet, from: .SetTime, to: .QueryReportControl)
        stateMachine.addTransition(transition: .GotReporterControl, from: .QueryReportControl, to: .QueryReporter0)
        stateMachine.addTransition(transition: .GotReporter0Attriburtes, from: .QueryReporter0, to: .QueryReporter1)
        stateMachine.addTransition(transition: .GotReporter1Attriburtes, from: .QueryReporter1, to: .QueryReporter2)
        stateMachine.addTransition(transition: .GotReporter2Attriburtes, from: .QueryReporter2, to: .Done)
    }
    
    func processResult(data: DataPacket) {
        debugPrint("state machine processing result \(stateMachine.state) --- \(data)")
        
        if stateMachine.state == DeviceLifecycle.QueryLock && data is LockPacket {
            let lock: LockPacket = data as! LockPacket
            
            debugPrint("QueryLock is what is going on \(lock.lock)")
            
            if lock.lock == 0 {
                _ = stateMachine.advance(transition: DeviceTransition.NeedLedCode)
                ledCodeNeeded = true
            } else {
                _ = stateMachine.advance(transition: DeviceTransition.GotLock)
            }
            
            debugPrint("state on exit of QueryLock is \(stateMachine.state)")
            
            return
        }
        
        if stateMachine.state.rawValue.lowercased().hasPrefix("sendledcode") {
            debugPrint("led code received, moving on")
            _ = stateMachine.advance(transition: stateMachine.state.getTransition())
            return
        }
        
        if stateMachine.state == DeviceLifecycle.QueryTime && data is TimePacket {
            debugPrint("Got time")
            let time: TimePacket = data as! TimePacket
            deviceTime = time.timestamp * 1000
            
          if abs(time.received.timeIntervalSince1970 - time.timestamp) < 60 {
                debugPrint("Time is correct")
               timeIsCorrect = true
            } else {
                timeDiff = time.received.timeIntervalSince1970 - time.timestamp
                debugPrint("time wrong timeDiff = \(timeDiff)")
            }
        }
        
        if stateMachine.state == DeviceLifecycle.QueryReporter0 && data is ReportAttributes {
            debugPrint("reporter 0 is right")
            reporter0Config = data as? ReportAttributes
        }
        
        if stateMachine.state == DeviceLifecycle.QueryReporter1 && data is ReportAttributes {
            debugPrint("reporter 1 is right")
            reporter1Config = data as? ReportAttributes
        }
        
        if stateMachine.state == DeviceLifecycle.QueryReporter2 && data is ReportAttributes {
            debugPrint("reporter2 is right")
            reporter0Config = data as? ReportAttributes
        }
        
        if stateMachine.state == DeviceLifecycle.QueryReportControl {
            let reportControl = data.reportControl
            debugPrint("looking at reportControl \(reportControl)")
            
            if reportControl & 0x01 == 0x01 {
                debugPrint("reporter 0 is on")
                reporter0On = true
            }
            
            if reportControl & 0x02 == 0x02 {
                debugPrint("reporter 1 is on")
                reporter1On = true
            }
            
            if reportControl & 0x04 == 0x04 {
                debugPrint("reporter 2 is on")
                reporter2On = true
            }
        }
        
        _ = stateMachine.advance(transition: stateMachine.state.getTransition())
        
        if stateMachine.state == DeviceLifecycle.SetTime && timeIsCorrect {
            debugPrint("time is correct next state please")
            _ = stateMachine.advance(transition: stateMachine.state.getTransition())
        }
        
        debugPrint("state on exit of processResult is \(stateMachine.state)")
    }
    
    func disconnected() {
        _ = stateMachine.advance(transition: DeviceTransition.Disconnected)
    }
    
    func reset() {
        stateMachine = StateMachine<DeviceLifecycle, DeviceTransition>(state: .QueryLock)
        stateMachine.addTransition(transition: .NeedLedCode, from: .QueryLock, to: .SendLedCode1)
        stateMachine.addTransition(transition: .GotLock, from: .QueryLock, to: .QueryTime)
        stateMachine.addTransition(transition: .GotTime, from: .QueryTime, to: .SetTime)
        stateMachine.addTransition(transition: .CodeWasReceived, from: .SendLedCode1, to: .SendLedCode2)
        stateMachine.addTransition(transition: .CodeWasReceived, from: .SendLedCode2, to: .SendLedCode3)
        stateMachine.addTransition(transition: .CodeWasReceived, from: .SendLedCode3, to: .SendLedCode4)
        stateMachine.addTransition(transition: .CodeWasReceived, from: .SendLedCode4, to: .QueryTime)
        stateMachine.addTransition(transition: .Disconnected, from: .SendLedCode4, to: .QueryLock)
        stateMachine.addTransition(transition: .TimeWasSet, from: .SetTime, to: .QueryReportControl)
        stateMachine.addTransition(transition: .GotReporterControl, from: .QueryReportControl, to: .QueryReporter0)
        stateMachine.addTransition(transition: .GotReporter0Attriburtes, from: .QueryReporter0, to: .QueryReporter1)
        stateMachine.addTransition(transition: .GotReporter1Attriburtes, from: .QueryReporter1, to: .QueryReporter2)
        stateMachine.addTransition(transition: .GotReporter2Attriburtes, from: .QueryReporter2, to: .Done)
    }
    
    func getState() -> DeviceLifecycle {
        return stateMachine.state
    }
    
    func isTimeCorrect() -> Bool {
        return timeIsCorrect
    }
    
    func isLedCodeNeeded() -> Bool {
        return ledCodeNeeded
    }
    
    func getDeviceTime() -> Double {
        return deviceTime
    }
}
